compiler/rustc_data_structures/src/functor.rs

```diff
@@ -0,0 +1,67 @@
+use rustc_index::vec::{Idx, IndexVec};
+use std::mem;
+use std::ptr;
+
+pub trait IdFunctor {
+    type Inner;
+
+    fn map_id<F>(self, f: F) -> Self
+    where
+        F: FnMut(Self::Inner) -> Self::Inner;
+}
+
+impl<T> IdFunctor for Box<T> {
+    type Inner = T;
+
+    fn map_id<F>(self, mut f: F) -> Self
+    where
+        F: FnMut(Self::Inner) -> Self::Inner,
+    {
+        let raw = Box::into_raw(self);
+        unsafe {
+            // SAFETY: The raw pointer points to a valid value of type `T`.
+            let value = ptr::read(raw);
+            // SAFETY: Converts `Box<T>` to `Box<MaybeUninit<T>>` which is the
+            // inverse of `Box::assume_init()` and should be safe.
+            let mut raw: Box<mem::MaybeUninit<T>> = Box::from_raw(raw.cast());
+            // SAFETY: Write the mapped value back into the `Box`.
+            ptr::write(raw.as_mut_ptr(), f(value));
+            // SAFETY: We just initialized `raw`.
+            raw.assume_init()
+        }
+    }
+}
+
+impl<T> IdFunctor for Vec<T> {
+    type Inner = T;
+
+    fn map_id<F>(self, f: F) -> Self
+    where
+        F: FnMut(Self::Inner) -> Self::Inner,
+    {
+        // FIXME: Assert that this reuses the old vec.
+        self.into_iter().map(f).collect()
```

Comment:
This does not seem to reuse the old vec, which is probably at least partially responsible for the slowdown